utilities/blob_db/blob_compaction_filter.cc

```diff
@@ -11,89 +11,273 @@
 namespace rocksdb {
 namespace blob_db {
 
-namespace {
+CompactionFilter::Decision BlobIndexCompactionFilterBase::FilterV2(
+    int /*level*/, const Slice& key, ValueType value_type, const Slice& value,
+    std::string* /*new_value*/, std::string* /*skip_until*/) const {
+  if (value_type != kBlobIndex) {
+    return Decision::kKeep;
+  }
+  BlobIndex blob_index;
+  Status s = blob_index.DecodeFrom(value);
+  if (!s.ok()) {
+    // Unable to decode blob index. Keeping the value.
+    return Decision::kKeep;
+  }
+  if (blob_index.HasTTL() && blob_index.expiration() <= current_time_) {
+    // Expired
+    expired_count_++;
+    expired_size_ += key.size() + value.size();
+    return Decision::kRemove;
+  }
+  if (!blob_index.IsInlined() &&
+      blob_index.file_number() < context_.next_file_number &&
+      context_.current_blob_files.count(blob_index.file_number()) == 0) {
+    // Corresponding blob file gone. Could have been garbage collected or
+    // evicted by FIFO eviction.
+    evicted_count_++;
+    evicted_size_ += key.size() + value.size();
+    return Decision::kRemove;
+  }
+  if (context_.fifo_eviction_seq > 0 && blob_index.HasTTL() &&
+      blob_index.expiration() < context_.evict_expiration_up_to) {
+    // Hack: Internal key is passed to BlobIndexCompactionFilter for it to
+    // get sequence number.
+    ParsedInternalKey ikey;
+    bool ok = ParseInternalKey(key, &ikey);
+    // Remove keys that could have been remove by last FIFO eviction.
+    // If get error while parsing key, ignore and continue.
+    if (ok && ikey.sequence < context_.fifo_eviction_seq) {
+      evicted_count_++;
+      evicted_size_ += key.size() + value.size();
+      return Decision::kRemove;
+    }
+  }
+  return Decision::kKeep;
+}
 
-// CompactionFilter to delete expired blob index from base DB.
-class BlobIndexCompactionFilter : public CompactionFilter {
- public:
-  BlobIndexCompactionFilter(BlobCompactionContext context,
-                            uint64_t current_time, Statistics* statistics)
-      : context_(context),
-        current_time_(current_time),
-        statistics_(statistics) {}
+bool BlobIndexCompactionFilterGC::PrepareBlobOutput(
```

Comment:
I just realized one thing: RocksDB only allows one compaction. So if a user uses BlobDB, then they cannot use their own compaction filter. It's OK for now but in the long term, maybe we should solve this problem.

Reply:
You mean one compaction filter, right?

Yes, BlobDB currently does not support (user-defined) compaction filters. There is actually a check to that effect in `Open`. It's part of the feature gap we need to close.